src/callbacks_step/analysis_dg2d_parallel.jl

```diff
@@ -12,53 +12,57 @@ function calc_error_norms(func, u, t, analyzer,
   # TODO: MPI, we should improve this; maybe we should dispatch on `u`
   #       and create some MPI array type, overloading broadcasting and mapreduce etc.
   #       Then, this specific array type should also work well with DiffEq etc.
-  l2_error, linf_error = invoke(calc_error_norms,
+  l2_errors, linf_errors = invoke(calc_error_norms_per_element,
     Tuple{typeof(func), typeof(u), typeof(t), typeof(analyzer), TreeMesh{2},
           typeof(equations), typeof(initial_condition), typeof(dg), typeof(cache),
           typeof(cache_analysis)},
     func, u, t, analyzer, mesh, equations, initial_condition, dg, cache, cache_analysis)
 
-  # Since the local L2 norm is already normalized and square-rooted, we need to undo this first
-  total_volume = mesh.tree.length_level_0^ndims(mesh)
-  global_l2_error = Vector(l2_error.^2 .* total_volume)
-  global_linf_error = Vector(linf_error)
-  MPI.Reduce!(global_l2_error, +, mpi_root(), mpi_comm())
-  MPI.Reduce!(global_linf_error, max, mpi_root(), mpi_comm())
+  nvars = length(l2_errors[1])
+  T = typeof(l2_errors[1]) # for type conversion of final errors
```

Comment:
Why not something like
```suggestion
  T = eltype(l2_errors)
```
?

Reply:
Changed it